Zend/zend_compile.c

```diff
@@ -7665,7 +7665,14 @@ static zend_string *zend_begin_func_decl(znode *result, zend_op_array *op_array,
 	zend_string *unqualified_name, *name, *lcname;
 	zend_op *opline;
 
-	unqualified_name = decl->name;
+	if (op_array->fn_flags & ZEND_ACC_CLOSURE) {
+		zend_string *filename = op_array->filename;
+		uint32_t start_lineno = decl->start_lineno;
+		unqualified_name = zend_strpprintf(0, "{closure:%s:%" PRIu32 "}", ZSTR_VAL(filename), start_lineno);
```

Comment:
Is this something that may just be emitted for stack traces, instead of changing the closure name? Otherwise, this breaks de-duplication of file names for interned strings, storing the file name many many times. They can be very long too.

Reply:
Having the information readily available can also be useful for debuggers and tracers as outlined by my mention of Xdebug in the PR description and the adjustments to the phpdbg and observer tests.

How would this be different from including the filename for anonymous classes which is already done? Would truncating the filename to just the basename help? This would only slightly reduce the usefulness within the name, but could greatly reduce the memory consumption.

Reply:
Anonymous classes are arguably many factors less common than closures. The default interned string buffer is only 8MB, so this change may potentially fill a significant amount of this memory with repeating strings.

If Xdebug already solves this problem, is there an aspect of it that does not suffice? Similarly, it seems to me observers and phpdbg can access this information from C, and output it to the user.

Note that I'm not questioning the usefulness of displaying the information in general. This is very clearly a win in that regard.

Reply:
> The default interned string buffer is only 8MB, so this change may potentially fill a significant amount of this memory with repeating strings.

I don't know how it works, but is this not an issue with regular function and method names as well? I expect regular functions and methods to exceed the number of closures and to also have distinct names. Or is this more about the length of the resulting function name where truncating the path to the basename would help reduce the impact?

> If Xdebug already solves this problem, is there an aspect of it that does not suffice?

That's probably where the “readily available” bit comes into play. By natively storing the information it does not need to be reimplemented in every tool and it also avoids the redundant computation of the extended closure name in hot loops. It's not particularly expensive, but at will likely require repeated allocations of a `zend_string` (unless somehow cached, but then we're back to memory usage).

Reply:
> I don't know how it works, but is this not an issue with regular function and method names as well? I expect regular functions and methods to exceed the number of closures and to also have mostly distinct names. Or is this more about the length of the resulting function name where truncating the path to the basename would help reduce the impact?

Right, the base name for all the file names is not deduplicated in the interned string buffer, nor are namespaces, so there's already some redundancy. References to the same class/function are deduplicated. If a file contains hundreds of closures, the same name will be stored hundreds of time for just this file, one time for each closure.

Maybe this is a non-issue, it's just something to think about. Maybe we just need to bump the default interned string buffer size. 8MB is a fairly low amount. It would be good to know much this increases memory consumption for something like Symfony Demo, or Wordpress.

Reply:
I wonder, is there a way to distinguish between "anonymous" closures and ``foo(...)`` closures? As the latter case doesn't need the file name and could just use the regular function name, would this reduce the memory usage?

Reply:
> I wonder, is there a way to distinguish between "anonymous" closures and `foo(...)` closures?

AFAIK the latter are reliably detected by `ZEND_ACC_FAKE_CLOSURE`. They do not appear to go through `zend_begin_func_decl()` anyway, though. 

Printing the unqualified name for:
```
<?php

$f = function () {
};

$trace = $f();

$g = intdiv(...);

$g(1, 0);

```

Results in:

```
{closure:*snip*/php-src/test.php:3}

Fatal error: Uncaught DivisionByZeroError: Division by zero in *snip*/php-src/test.php:10
Stack trace:
#0 *snip*/php-src/test.php(10): intdiv(1, 0)
#1 {main}
  thrown in *snip*/php-src/test.php on line 10
```

Note how they do not show up in the stack trace either.

Reply:
> It would be good to know much this increases memory consumption for something like Symfony Demo, or Wordpress.

I've cloned https://github.com/php/benchmarking-symfony-demo-2.2.3 locally and then did the following:

With an `printf("%s\n", ZSTR_VAL(unqualified_name));` for closures and running `sapi/cgi/php-cgi public/index.php |grep '^{closure'` I get 284 names:

<details><summary>284 Closure names</summary>

```
{closure:/tmp/benchmarking-symfony-demo-2.2.3/public/index.php:9}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/composer/autoload_real.php:37}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/composer/ClassLoader.php:581}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/composer/autoload_static.php:654}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/public/index.php:9}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/dotenv/Dotenv.php:436}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/dotenv/Dotenv.php:486}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/runtime/GenericRuntime.php:90}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/runtime/GenericRuntime.php:116}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/console/Input/ArgvInput.php:350}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/console/Input/Input.php:65}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/Kernel.php:432}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/config/ResourceCheckerConfigCache.php:150}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:133}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:171}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:174}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/Request.php:567}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/Request.php:608}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/AcceptHeader.php:53}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/AcceptHeader.php:118}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/AcceptHeader.php:139}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/FileBag.php:78}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/AbstractAdapter.php:47}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/AbstractAdapter.php:60}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Traits/ContractsTrait.php:57}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Traits/ContractsTrait.php:73}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Traits/ContractsTrait.php:87}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Traits/ContractsTrait.php:104}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/ArrayAdapter.php:61}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/ChainAdapter.php:70}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/ChainAdapter.php:94}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/ChainAdapter.php:103}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/PhpArrayAdapter.php:53}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/PhpFilesAdapter.php:52}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/error-handler/ErrorHandler.php:164}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/error-handler/ErrorHandler.php:188}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/error-handler/ErrorHandler.php:636}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/error-handler/ErrorHandler.php:725}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:60}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:81}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:108}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:141}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:158}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:183}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:207}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Hydrator.php:245}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/doctrine-bundle/DoctrineBundle.php:93}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:137}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:140}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:143}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:146}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:149}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:152}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:155}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:158}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:161}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:164}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:167}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:170}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:173}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:176}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:179}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:182}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:185}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:188}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:191}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:194}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:197}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:200}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:203}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:206}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:209}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:212}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:215}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:218}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:221}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:224}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:227}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:230}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:233}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:236}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:239}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:242}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:245}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:248}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:251}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:254}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:257}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:260}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:263}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:266}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:269}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:272}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:275}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:278}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:281}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:284}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:287}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:290}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:293}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:296}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:299}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:302}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:305}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:308}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:311}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:331}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:501}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:658}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:689}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:723}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:726}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:729}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:732}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:735}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:738}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:741}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:744}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:747}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:750}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:753}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:756}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:759}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:762}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:765}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:768}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/App_KernelProdContainer.php:809}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/dependency-injection/Argument/ServiceLocator.php:46}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/Autoloader.php:77}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/event-dispatcher/EventDispatcher.php:241}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/Controller/ControllerResolver.php:200}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/monolog/monolog/src/Monolog/Utils.php:153}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/monolog/monolog/src/Monolog/Utils.php:202}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/monolog/monolog/src/Monolog/Formatter/JsonFormatter.php:137}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/monolog/monolog/src/Monolog/Handler/FingersCrossedHandler.php:185}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/EventListener/DebugHandlersListener.php:119}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/EventListener/DebugHandlersListener.php:133}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/EventListener/AbstractSessionListener.php:70}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/framework-bundle/Routing/Router.php:161}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Router.php:254}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Router.php:288}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Matcher/ExpressionLanguageProvider.php:39}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Matcher/ExpressionLanguageProvider.php:42}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/dependency-injection/ParameterBag/ParameterBag.php:193}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getContainer_EnvVarProcessorService.php:20}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/dependency-injection/EnvVarProcessor.php:298}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/Slugger/AsciiSlugger.php:124}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/translation/Translator.php:181}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/translation/Translator.php:258}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getTranslation_LocaleSwitcherService.php:20}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/Firewall.php:69}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-csrf/CsrfTokenManager.php:45}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-csrf/CsrfTokenManager.php:52}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getSecurity_Firewall_Map_Context_MainService.php:28}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/HttpUtils.php:77}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/doctrine-bridge/ManagerRegistry.php:58}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-bundle/Security/LazyFirewallContext.php:70}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/Firewall/ContextListener.php:258}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/Authentication/AuthenticatorManager.php:79}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/Authenticator/RememberMeAuthenticator.php:90}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/cache/Adapter/NullAdapter.php:28}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/framework-bundle/Controller/AbstractController.php:295}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/src/Twig/SourceCodeExtension.php:149}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/src/Twig/SourceCodeExtension.php:155}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/doctrine-bridge/ArgumentResolver/EntityValueResolver.php:189}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/expression-language/ExpressionFunction.php:85}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/expression-language/ExpressionFunction.php:89}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getDoctrine_Orm_DefaultEntityManagerService.php:21}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getDoctrine_Orm_DefaultEntityManagerService.php:21}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:70}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:79}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:92}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:94}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:102}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:105}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/LazyObjectRegistry.php:108}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/Configuration.php:63}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/cache/lib/Doctrine/Common/Cache/Psr6/DoctrineProvider.php:39}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/DefaultQuoteStrategy.php:123}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/deprecations/lib/Doctrine/Deprecations/Deprecation.php:252}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/Schema/SqliteSchemaManager.php:241}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:2277}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:2321}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:2366}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:2410}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:2470}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:3473}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:3483}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Proxy/ProxyFactory.php:153}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Proxy/ProxyFactory.php:180}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Proxy/ProxyFactory.php:229}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Proxy/ProxyFactory.php:266}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:431}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:489}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:1164}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:1179}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:1226}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/common/src/Proxy/ProxyGenerator.php:1239}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/src/Repository/PostRepository.php:103}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/doctrine-bundle/Repository/LazyServiceEntityRepository.php:44}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/EventListener/IsGrantedAttributeListener.php:103}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-http/EventListener/IsGrantedAttributeListener.php:108}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/EventListener/CacheAttributeListener.php:77}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/doctrine/orm/default_metadata.php:14}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/doctrine/orm/default_metadata.php:142}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/doctrine/orm/default_metadata.php:383}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/doctrine/orm/default_metadata.php:548}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:2314}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:3676}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/instantiator/src/Doctrine/Instantiator/Instantiator.php:135}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/instantiator/src/Doctrine/Instantiator/Instantiator.php:178}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/persistence/src/Persistence/Reflection/TypedNoDefaultReflectionPropertyBase.php:54}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/QueryBuilder.php:643}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:333}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:514}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:1189}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:1207}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:1407}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/Types/Type.php:204}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/Types/ArrayType.php:53}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/Types/ObjectType.php:53}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/pools/system/5Rb4vK0BNn/T/M/y4pd5KFnniihz63+n6mg:3}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getDoctrine_Orm_Listeners_DoctrineTokenProviderSchemaSubscriberService.php:20}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:636}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:704}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Tools/Pagination/WhereInWalker.php:149}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/pools/system/5Rb4vK0BNn/Z/N/PCAqz2GzHgHf0cy5hx-g:3}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1959}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:2148}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/var-exporter/Internal/Exporter.php:222}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/SQL/Parser.php:84}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/SQL/Parser.php:87}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/SQL/Parser.php:90}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/dbal/src/SQL/Parser.php:93}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/pools/system/5Rb4vK0BNn/L/S/vRLrvPI5h4y8NmpV3OCA:3}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Extension/CoreExtension.php:1032}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Extension/CoreExtension.php:1568}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Extension/EscaperExtension.php:258}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Extension/EscaperExtension.php:309}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Extension/EscaperExtension.php:330}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/var/cache/prod/ContainerTkK9Fxx/getAssets_PackagesService.php:22}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/twig-bridge/Extension/CodeExtension.php:123}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/twig-bridge/Extension/CodeExtension.php:191}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/doctrine/doctrine-bundle/Twig/DoctrineExtension.php:120}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/ux-twig-component/src/Twig/ComponentExtension.php:51}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/twig-bundle/DependencyInjection/Configurator/EnvironmentConfigurator.php:56}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/twig-bundle/DependencyInjection/Configurator/EnvironmentConfigurator.php:57}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/TemplateWrapper.php:70}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Template.php:216}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Template.php:241}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/twig/twig/src/Template.php:376}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Generator/UrlGenerator.php:272}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/routing/Generator/UrlGenerator.php:276}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/webpack-encore-bundle/src/Asset/TagRenderer.php:44}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/webpack-encore-bundle/src/Asset/TagRenderer.php:181}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/webpack-encore-bundle/src/Asset/TagRenderer.php:186}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/webpack-encore-bundle/src/Asset/TagRenderer.php:186}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-foundation/Session/Storage/NativeSessionStorage.php:240}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:27}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:29}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:33}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:35}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:39}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:41}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:45}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/security-core/Authorization/ExpressionLanguageProvider.php:47}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-mX-Powered-By: PHP/8.4.0-dev
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-mbstring/Mbstring.php:183}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-icu/DateFormat/FullTransformer.php:79}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-icu/DateFormat/FullTransformer.php:166}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-icu/DateFormat/FullTransformer.php:215}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-icu/DateFormat/FullTransformer.php:215}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-icu/DateFormat/MonthTransformer.php:56}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/http-kernel/Fragment/InlineFragmentRenderer.php:121}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:142}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:162}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:162}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:233}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:302}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:317}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractUnicodeString.php:363}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractString.php:451}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/string/AbstractString.php:510}
{closure:/tmp/benchmarking-symfony-demo-2.2.3/vendor/symfony/polyfill-intl-grapheme/Grapheme.php:53}
```
</details>

I then removed the `printf()` and added the following to `public/index.php`:

```
register_shutdown_function(function () {
        var_dump(opcache_get_status(false));
});
```

Then I used the following command:

```
*snip*/php-src/sapi/cgi/php-cgi -dzend_extension=*snip*/php-src/modules/opcache.so -dop
cache.enable=1 public/index.php
```

With this PR applied:

```
array(8) {
  ["opcache_enabled"]=>
  bool(true)
  ["cache_full"]=>
  bool(false)
  ["restart_pending"]=>
  bool(false)
  ["restart_in_progress"]=>
  bool(false)
  ["memory_usage"]=>
  array(4) {
    ["used_memory"]=>
    int(22858016)
    ["free_memory"]=>
    int(111351560)
    ["wasted_memory"]=>
    int(8152)
    ["current_wasted_percentage"]=>
    float(0.006073713302612305)
  }
  ["interned_strings_usage"]=>
  array(4) {
    ["buffer_size"]=>
    int(8388608)
    ["used_memory"]=>
    int(5074032)
    ["free_memory"]=>
    int(3314576)
    ["number_of_strings"]=>
    int(48739)
  }
  ["opcache_statistics"]=>
  array(13) {
    ["num_cached_scripts"]=>
    int(836)
    ["num_cached_keys"]=>
    int(1562)
    ["max_cached_keys"]=>
    int(16229)
    ["hits"]=>
    int(1)
    ["start_time"]=>
    int(1709133905)
    ["last_restart_time"]=>
    int(0)
    ["oom_restarts"]=>
    int(0)
    ["hash_restarts"]=>
    int(0)
    ["manual_restarts"]=>
    int(0)
    ["misses"]=>
    int(837)
    ["blacklist_misses"]=>
    int(0)
    ["blacklist_miss_ratio"]=>
    float(0)
    ["opcache_hit_rate"]=>
    float(0.11933174224343676)
  }
  ["jit"]=>
  array(7) {
    ["enabled"]=>
    bool(false)
    ["on"]=>
    bool(false)
    ["kind"]=>
    int(0)
    ["opt_level"]=>
    int(0)
    ["opt_flags"]=>
    int(0)
    ["buffer_size"]=>
    int(0)
    ["buffer_free"]=>
    int(0)
  }
}
```

Without this PR applied:

```
array(8) {
  ["opcache_enabled"]=>
  bool(true)
  ["cache_full"]=>
  bool(false)
  ["restart_pending"]=>
  bool(false)
  ["restart_in_progress"]=>
  bool(false)
  ["memory_usage"]=>
  array(4) {
    ["used_memory"]=>
    int(22858016)
    ["free_memory"]=>
    int(111351560)
    ["wasted_memory"]=>
    int(8152)
    ["current_wasted_percentage"]=>
    float(0.006073713302612305)
  }
  ["interned_strings_usage"]=>
  array(4) {
    ["buffer_size"]=>
    int(8388608)
    ["used_memory"]=>
    int(5033520)
    ["free_memory"]=>
    int(3355088)
    ["number_of_strings"]=>
    int(48530)
  }
  ["opcache_statistics"]=>
  array(13) {
    ["num_cached_scripts"]=>
    int(836)
    ["num_cached_keys"]=>
    int(1562)
    ["max_cached_keys"]=>
    int(16229)
    ["hits"]=>
    int(1)
    ["start_time"]=>
    int(1709133950)
    ["last_restart_time"]=>
    int(0)
    ["oom_restarts"]=>
    int(0)
    ["hash_restarts"]=>
    int(0)
    ["manual_restarts"]=>
    int(0)
    ["misses"]=>
    int(837)
    ["blacklist_misses"]=>
    int(0)
    ["blacklist_miss_ratio"]=>
    float(0)
    ["opcache_hit_rate"]=>
    float(0.11933174224343676)
  }
  ["jit"]=>
  array(7) {
    ["enabled"]=>
    bool(false)
    ["on"]=>
    bool(false)
    ["kind"]=>
    int(0)
    ["opt_level"]=>
    int(0)
    ["opt_flags"]=>
    int(0)
    ["buffer_size"]=>
    int(0)
    ["buffer_free"]=>
    int(0)
  }
}
```

That's a difference of 48739-48530 = 209 additional interned strings and 5074032-5033520 = 40512 additional bytes of interned string memory usage (roughly 193 byte per interned string and 142 byte per closure).

Truncating the included filename to just the basename via:

```
		char *slash;
		if ((slash = memrchr(ZSTR_VAL(filename), '/', ZSTR_LEN(filename))) == NULL) {
			slash = ZSTR_VAL(filename);
		}
		unqualified_name = zend_strpprintf(0, "{closure:%s:%" PRIu32 "}", slash, start_lineno);
```

Does not change the number of interned strings, but brings the memory usage down to 5054040 bytes, which is 20520 bytes additional compared to the state without this PR, 98 bytes per interned string and 72 bytes per closure.

Reply:
Thank you for checking! That's roughly +0.8%, which indeed doesn't sound too bad. Of course, it's hard to say whether this is representative, but at least we have a rough idea of the increase now.

Reply:
I'm also not sure, why do we need to keep the names for closures.
We may always reconstruct them, concatenating "filename" and "start_lineno" when necessary (in zend_fetch_debug_backtrace).

Reply:
> We may always reconstruct them, concatenating "filename" and "start_lineno" when necessary (in zend_fetch_debug_backtrace).

The “full” name of the function is not just useful for backtraces. This PR also adds the name to the `__debugInfo()` output of closures, allowing users to easily see the closure origin when dumping a variable containing it. A reusable function for closure naming could be provided in php-src, but that would still require everyone to use it to achieve consistent naming results for the developer. By including it in the function name field there would be no inconsistencies. The additional function calls in various locations would also increase machine code size.

See also the second paragraph of my previous comment: https://github.com/php/php-src/pull/13550#discussion_r1506071825

-----------------

The question is whether the time-memory-tradeoff is worth it – both from a CPU time (for the repeated recomputation) and from a developer time (for the improved consistency) perspective.

Reply:
In production use we should rarely have calls to ``debug_backtrace`` and ``__debugInfo()``, so that closure names are going to be never used (just kept in the memory). Both ``debug_backtrace`` and ``__debugInfo()`` already construct arrays and the additional name reconstruction won't make a big difference. 

I think moving this name reconstruction code from ``zend_compile.c`` into ``zend_closures.c`` will only make the thing more clear. This will also remove the new ``zend_string_copy()/zend_string_release()`` calls that you introduced there for each op_array.

Reply:
I've realized that adding the filename to the function name made the namespace effectively redundant. As Closures previously had a static “unqualified name” there should be no risk of collision either. Dropping the namespace reduced the total memory usage with the Symfony demo to 

```
  ["interned_strings_usage"]=>
  array(4) {
    ["buffer_size"]=>
    int(8388608)
    ["used_memory"]=>
    int(5067336)
    ["free_memory"]=>
    int(3321272)
    ["number_of_strings"]=>
    int(48739)
  }
```
which is 209 additional interned strings and 33816 bytes of additional memory usage. Truncating the filename to just the basename would result in 13840 bytes of additional memory usage (but just the basename without the namespace might be too unspecific).

This change also allowed me to simplify the implementation in `zend_compile.c` to just an additional `if (op_array->fn_flags & ZEND_ACC_CLOSURE) {`, without touching the logic for regular functions at all.

What do you think?